plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,424 @@
+                .inputs(buildInputs(queryCompletedEvent.getMetadata()))
+                .outputs(buildOutputs(queryCompletedEvent.getIoMetadata()))
+                .build();
+    }
+
+    private RunFacetsBuilder getBaseRunFacetsBuilder(QueryContext queryContext)
+    {
+        return openLineage.newRunFacetsBuilder()
+                .processing_engine(openLineage.newProcessingEngineRunFacetBuilder()
+                        .name("trino")
+                        .version(queryContext.getServerVersion())
+                        .build());
+    }
+
+    private JobBuilder getBaseJobBuilder(QueryMetadata queryMetadata)
+    {
+        return openLineage.newJobBuilder()
+                .namespace(this.jobNamespace)
+                .name(queryMetadata.getQueryId())
+                .facets(openLineage.newJobFacetsBuilder()
+                            .jobType(openLineage.newJobTypeJobFacet("BATCH", "TRINO", "QUERY"))
+                            .sql(openLineage.newSQLJobFacet(queryMetadata.getQuery()))
+                            .build());
+    }
+
+    private List<InputDataset> buildInputs(QueryMetadata queryMetadata)
+    {
+        HashMap<String, Set<TableInfo>> directlyReferencedToSymlinksMap = new HashMap<>();
+
+        queryMetadata
+                .getTables()
+                .stream()
+                .sorted(Comparator.comparing(TableInfo::isDirectlyReferenced, Comparator.reverseOrder()))
+                .forEach(table -> {
+                    if (table.isDirectlyReferenced()) {
+                        directlyReferencedToSymlinksMap.put(getDatasetName(table),
+                                new HashSet<>());
+                    }
+                    else {
+                        if (!table.getReferenceChain().isEmpty()) {
+                            table
+                                    .getReferenceChain()
+                                    .forEach(reference -> {
+                                        String referenceDatasetName = null;
+                                        if (reference instanceof BaseViewReferenceInfo) {
+                                            referenceDatasetName = getDatasetName(((BaseViewReferenceInfo) reference).catalogName(), ((BaseViewReferenceInfo) reference).schemaName(), ((BaseViewReferenceInfo) reference).viewName());
+                                        }
+                                        else if (reference instanceof FilterMaskReferenceInfo) {
+                                            referenceDatasetName = getDatasetName(((FilterMaskReferenceInfo) reference).targetCatalogName(), ((FilterMaskReferenceInfo) reference).targetSchemaName(), ((FilterMaskReferenceInfo) reference).targetTableName());
+                                        }
+
+                                        if (referenceDatasetName != null) {
+                                            directlyReferencedToSymlinksMap.get(referenceDatasetName).add(table);
```

Comment:
What if we have an intermediate view ? i.e

CREATE TABLE table1;
CREATE VIEW view1 AS SELECT * FROM table1;
CREATE VIEW view2 AS SELECT * FROM view1;

CREATE TABLE new_table AS SELECT * FROM view2;

We could get a `view1` referenced but this map would throw a NPE.

Reply:
ok, so I propose to:
- catch this NPE and skip on error
- extend `include-query-type`s with `QueryType.DATA_DEFINITION`

in such case, when view is created we register this information, and then in case of the lineage scenario you brought up we will have 2 symlink entities (table and view) and they both will be already registered in OL

Reply:
ok, so as discussed - removed symlinks and rely on directly referenced entities from getTables only + included QueryType.DATA_DEFINITION so we capture view creation as well